src/io/flutter/run/LaunchState.java

```diff
@@ -326,6 +328,21 @@ protected final RunContentDescriptor doExecute(@NotNull RunProfileState state, @
 
         if (app != null && StringUtil.equals(app.deviceId(), selectedDeviceId)) {
           if (executorId.equals(app.getMode().mode())) {
+
+            if (!identicalCommands(app.getCommand(), launchState.runConfig.getCommand(env, app.device()))) {
+              // To be safe, relaunch as the arguments to launch have changed.
+              try {
+                // TODO(jacobr): ideally we shouldn't be synchronously waiting
+                // for futures like this but I don't see a better option.
+                // In practice this seems fine.
+                app.shutdownAsync().get();
```

Comment:
Thinking about this, you may not actually need to shut down the old launch; launching the new one will cause the old one to be terminated by flutter_tools.

Reply:
I'm surprised the shutdown call is required but it is currently.
unfortunately the shutdown is required as otherwise we surprisingly end up with some stale state. Even if it wasn't required, clarifying the behavior by shutting down the application immediately is probably better as otherwise you would get 10 confusing seconds with an existing old app running even though a rebuild is in progress.